vm/src/dictdatatype.rs

```diff
@@ -0,0 +1,216 @@
+            if let Some(entry) = &self.entries[index] {
+                Ok(entry.value.clone())
+            } else {
+                panic!("Lookup returned invalied index into entries!");
+            }
+        } else {
+            let key_repr = vm.to_pystr(key)?;
+            Err(vm.new_value_error(format!("Key not found: {}", key_repr)))
+        }
+    }
+
+    /// Delete a key
+    pub fn delete(
+        &mut self,
+        vm: &mut VirtualMachine,
+        key: &PyObjectRef,
+    ) -> Result<(), PyObjectRef> {
+        if let LookupResult::Existing(index) = self.lookup(vm, key)? {
+            self.entries[index] = None;
+            Ok(())
+        } else {
+            let key_repr = vm.to_pystr(key)?;
+            Err(vm.new_value_error(format!("Key not found: {}", key_repr)))
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.get_items().len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn get_items(&self) -> Vec<(PyObjectRef, PyObjectRef)> {
+        self.entries
+            .iter()
+            .filter(|e| e.is_some())
+            .map(|e| e.as_ref().unwrap())
+            .map(|e| (e.key.clone(), e.value.clone()))
+            .collect()
+    }
+
+    /// Lookup the index for the given key.
+    fn lookup(
+        &self,
+        vm: &mut VirtualMachine,
+        key: &PyObjectRef,
+    ) -> Result<LookupResult, PyObjectRef> {
+        let hash_value = calc_hash(vm, key)?;
+        let perturb = hash_value;
+        let mut i: usize = hash_value;
+        loop {
+            if self.indices.contains_key(&i) {
+                // Now we have an index, lets check the key.
+                let index = self.indices[&i];
+                if let Some(entry) = &self.entries[index] {
+                    // Okay, we have an entry at this place
+                    if entry.key.is(key) {
+                        // Literally the same object
+                        break Ok(LookupResult::Existing(index));
+                    } else if entry.hash == hash_value {
+                        if do_eq(vm, &entry.key, key)? {
+                            break Ok(LookupResult::Existing(index));
+                        } else {
+                            // entry mismatch.
+                        }
+                    } else {
+                        // entry mismatch.
+                    }
+                } else {
+                    // Removed entry, continue search...
+                }
+            } else {
+                // Hash not in table, we are at free slot now.
+                break Ok(LookupResult::NewIndex { hash_value, i });
+            }
+
+            // Update i to next probe location:
+            i = i.wrapping_mul(5).wrapping_add(perturb).wrapping_add(1);
+            warn!("Perturb value: {}", i);
+        }
+    }
+}
+
+enum LookupResult {
+    NewIndex { hash_value: usize, i: usize }, // return not found, index into indices
```

Comment:
Should this be ```index``` rather than ```i```?

Reply:
No, to avoid confusion, it should be named not index. It is the index into the first mapping table. I'll call it `hash_index`.

Reply:
And I was indeed confused. :)